### PR TITLE
[lora] Fix WAR race: never write MoE runner output into hidden_states in place

### PR DESCRIPTION
**Pull request by sshleifer**
## Motivation

- We have a long decode lora eval that improves with this fix. Generates some occasional gibberish without.

## Root cause

A write-after-read race between the LoRA MoE runner and dual-stream shared experts:

- With LoRA, the marlin MoE path writes its final reduction **into `hidden_states`**: `output = hidden_states if inplace else torch.empty_like(hidden_states)` with `MoeRunnerConfig.inplace=True` by default. The fused non-LoRA kernels allocate their own output buffer — only the LoRA path aliases, which is why base-model decode is unaffected.
- `DeepseekV2MoE.forward_normal_dual_stream` (taken during decode cuda-graph capture) reads `hidden_states` for the shared experts on the alt stream. The only fork edge (`alt_stream.wait_stream(current_stream)`) is recorded before the router, so there is no ordering between the routed-experts write and the shared-expert read. The in-code comment documents a read-only assumption ("the routed deep_gemm does not free hidden_states, which the shared expert reads on the alt stream") — disposal is handled, mutation is not.
- The eager path already knows about this contract: `forward_normal` orders shared-vs-routed off the flag (`defer_shared = not self.experts.moe_runner_config.inplace`). The dual-stream path has no equivalent handling.

Inside the captured decode graph, whether the alt-stream read beats the main-stream write is scheduling luck. When it loses, the shared expert consumes routed-output rows instead of layer input — subtly corrupted logits every affected step, compounding through the KV cache. This matches all observed properties: LoRA-only, stochastic per run, variable onset, tail-worst, text coherent with non-canonical token boundaries.

## Modifications

- Force `inplace=False` at `FusedMoEWithLoRA` construction, mirroring the existing FlashInfer TRTLLM force-disable in `fused_moe_triton/layer.py`. 
- This makes the LoRA runner match the allocation behavior of every fused non-LoRA path and covers all models with dual-stream MoE forwards in one place. 

## Validation

3072-token forced-greedy probe, scored by re-tokenizing the generated text (canonical ≈ 3070/3072; fragmented generations re-tokenize far lower). GB300 tp4, Kimi-K2.5 int4 marlin + MoE LoRA, 8 reps per server:

| arm | reps | collapsed (<90% canonical) | min |
|---|---|---|---|
| inplace write active (2 servers + field runs) | 16 | 4 | 2332/3072 |
| write removed | 24 | 0 (nothing below 2700) | 2992/3072 |

Steady-state decode throughput and tpot are unchanged by the fix.

---
### CI States

Latest PR Test (Base): :x: [Run #30075718509](https://github.com/sgl-project/sglang/actions/runs/30075718509)
Latest PR Test (Extra): :x: [Run #30076552439](https://github.com/sgl-project/sglang/actions/runs/30076552439)

### CODE REVIEW

**Comment by sshleifer** [comment]
Deterministic minimal repro of the race (single GPU, no serving stack): replicate forward_normal_dual_stream's exact stream edges (alt_stream.wait_stream(cur) recorded before the routed enqueue; shared read under torch.cuda.stream(alt); join after) with the routed step writing into hidden_states per fused_marlin_moe's `output = hidden_states if inplace else torch.empty_like(...)`, then compare to the eager reference.

```
inplace=True  delay=0     max|out-ref| = 6.94   CORRUPTED (== analytic W(x)+V(W(x)) exactly)
inplace=True  delay=BIG   max|out-ref| = 6.94   CORRUPTED (race amplified via torch.cuda._sleep on alt stream)
inplace=False delay=BIG   max|out-ref| = 0.00   CORRECT under the same hostile schedule
```

Since a delay kernel can't change the output of a correctly synchronized program (it only realizes schedules the dependency edges already permit), the mismatch proves the WAR race; the corruption being exactly 'shared expert consumed the routed output' confirms the mechanism this PR fixes.

**Comment by sshleifer** [comment]
Added a CPU-only regression test to this PR: it runs the real `FusedMoEWithLoRA.__init__` against a config constructed with explicit `inplace=True` (as `FusedMoE.__init__` does) and asserts the shared config object comes out `inplace=False` for both the triton-fallback and marlin lora runner cores. Mutation-checked: reverting the one-line fix fails both tests.

**Review by yushengsu-thu — APPROVED** [review]
